### ISSUE TITLE
Training Fails with attn_implementation="flash_attention_2" in gemma-2-9b Model

### ISSUE THREAD

**Issue by BlackNoodle** [huggingface/transformers#31997]
### System Info

```
torch==2.3.1
flash-attn==2.6.1
transformers @ git+https://github.com/huggingface/transformers.git@ac946aac257cadfa8264fa4a284cd0ea1061c5b5
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

When training a model using the transformers library’s Trainer, the first step logs the loss and learning rate correctly, but the grad_norm appears as nan. In the subsequent steps, the loss shows as 0, grad_norm remains nan, and training does not proceed.

Model Loading Code:
```
from transformers import AutoModelForCausalLM, Trainer

model = AutoModelForCausalLM.from_pretrained(
    "google/gemma-2-9b",
    attn_implementation="flash_attention_2",
    use_cache=False,
    trust_remote_code=True,
    torch_dtype=torch.bfloat16,
)
```

First Step Log:
```
{
    'loss': 2.3101,
    'grad_norm': nan,
    'learning_rate': 4.000000000000001e-06,
    'gpu_allocated': 17644.15234375,
    'gpu_max_allocated': 54432.4921875,
    'gpu_reserved': 68638.75,
    'gpu_max_reserved': 78825.0,
    'epoch': 0.03
}
```

Subsequent Steps Log:
```
{
    'loss': 0.0,
    'grad_norm': nan,
    'learning_rate': 8.000000000000001e-06,
    'gpu_allocated': 17644.15234375,
    'gpu_max_allocated': 63246.1171875,
    'gpu_reserved': 37002.0,
    'gpu_max_reserved': 79341.25,
    'epoch': 0.06
}
```

Additional Information:
When the attn_implementation parameter is changed to 'eager' with the same code, data, and model, training proceeds without any issues.

### Expected behavior

Training should proceed without any issues.

**Comment by amyeroberts** [comment]
Hi @BlackNoodle, thanks for raising an issue! 

Could you share a minimal code reproduce we can use to run and replicate these results? In particular showing how trainer is being setup. 

cc @muellerzr @ArthurZucker

**Comment by BlackNoodle** [comment]
I think it is caused by: 
https://github.com/Dao-AILab/flash-attention/issues/1065

**Comment by Masrkai** [comment]
Can confirm that the issue doesn't exist on my tests with future versions of the model `gemma3` and it's no longer relevant.

**Comment by tolgadur** [comment]
I am getting this warning with gemma3 on the grpotrainer https://github.com/huggingface/transformers/blob/main/src/transformers/models/gemma3/modeling_gemma3.py#L657-L661. can we ignore it if it's not relevant anymore?
